Added Grafana-agent page for the decision on those (#30)
distributed-technologies/SCCP-documentation: docs/decisions/0001-grafana-agents.md

> - Documentation is confusing

Reviewer: ```suggestion
# Grafana Agents - go/no go?

Reasons for ditching Grafana Agents:
- Documentation is confusing
```

- Agent doc and operator doc is not aligned

> - Doc makes working with this developer hostile

Author: The developers in our team?

What I was trying to say is that working with documentation that are not aligend and more confusing then any thing else is developer hostile, from my point of view?

> - Doc makes working with this developer hostile

Reviewer: Hmm, I see. The sentence isn't easy to read (IMO).

Could we just change:
```
- Documentation is confusing
- Agent doc and operator doc is not aligned
- Doc makes working with this developer hostile
```

to

```
- The documentation is confusing and the Agents and operator docs aren't aligned
```

> - Doc makes working with this developer hostile

Author: Works for me :+1 and sounds better

- Functionality we though we wanted to use is not pressent - integrations and traces
- In the current Agent configuration, we're using it as a package deal for promtail and prometheus.
- We send all logs to one place

> - We want all metrics to be scraped by one agent anyways.

Author: That the proposed solution that was build did not utilize the part where we could spin up Grafana agents anywhere, that setup just had one agent doing all the scraping

> - We want all metrics to be scraped by one agent anyways.

Reviewer: Per cluster or per node?

> - We want all metrics to be scraped by one agent anyways.

Author: The metrics scraping was done with one pr. cluster, the log scraping was one pr. node

> - We want all metrics to be scraped by one agent anyways.

Reviewer: Hmm, okay. I think this should be reworded and/or moved somewhere else in the document. It sounds more like a fact than a argument against Grafana Agents.
```
- In the current Agent configuration, we're using it as a package deal for promtail and prometheus.
- We send all logs to one place
- We want all metrics to be scraped by one agent anyways.
```

- Curently configuration options are lackluster

The reason we choose grafana agents is that we wanted a lightweight way of using the `remote write` feature to write to cortex, also this would allow us to get rid of a bulky Prometheus instance.

> <br>

Reviewer 2: Does this actually do anything in markdown?

Why should we change back to prometheus and promtail

> - Prometheus has an agent feature in experimental, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.

Reviewer: ```suggestion
- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale
```

## Grafana agents - go/no go?

Reasons for not going with Grafana agents
- The documentation is confusing and the Agents and operator docs aren't aligened.
- Some of the main sellingpoints (integrations and traces) aren't supported by the operator
- In the current Agent configuration, we're using it as a package deal for promtail and prometheus.
- We send all logs to one place
- We want all metrics to be scraped by one agent anyways.

> - Curently configuration options are lackluster

Reviewer: Anything is particular missing?

<br>

Why should we change back to prometheus and promtail
- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
 It sounds like it would achive the same goal of giving us a instance that can `remote write` to cortex, without storing metrics locally.
- Prometheus is more mature for the time being.
- The prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing higher configuration options.
  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
- For Promtail, I am currently seeing issues related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)

> 
> So from my point of view we should consider going back and use prometheus and promtail,
> 
> This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.
> 
> I would suggest using the prometheus operator for convinience and then deploying promtail seperate to Loki.

Reviewer: ```suggestion

### Recommendation

So from my point of view we should consider going back and use Prometheus and promtail,

This is mainly due to the inconsistent doc and early stage of the Grafana Agents, and we need something that is more mature and has more consistent doc.

I would suggest using the prometheus operator for convenience and then deploying promtail separate to Loki.
```

Author: For the flexibility, I suppose either way if we want to use the distributed version of Loki, then we would have to build a dependency into our proxy chart.

But mostly the flexibility?

> ## Grafana agents - go/no go?
> 
> Reasons for not going with Grafana agents
> - The documentation is confusing and the Agents and operator docs aren't aligened.
> - Some of the main selling points (integrations and traces) aren't supported by the operator
> - Generally there were less configuration options in the Grafana Agent Operators.
>   - One example is the implementation of Promtail: The `fs.inotify.max_user_instances` value can be changed in Promtails chart, but is not mentioned in the Grafana Agents implementation.
>   - It is also hard to make changes directly to one of the pods using arguments, you cannot add a argurment, you have to provide **all** the arguments if you want to add or change one.
> 
> Grafana Agents was chosen initially as a way to avoid a builky Prometheus instance, as we only need the `remote write` functionality, so we can send metrics to Cortex.

Reviewer 2: ```suggestion
## Grafana agents - Are not mature

Reasons for not going with Grafana agents
- The documentation is confusing and the Agents and operator docs are not aligned.
- Some of the main selling points (integrations and traces) are not supported by the operator
- Generally there were less configuration options in the Grafana Agent Operators.
  - One example is the implementation of Promtail: The `fs.inotify.max_user_instances` value can be changed in Promtails chart, but is not mentioned in the Grafana Agents implementation.
  - It is also hard to make changes directly to one of the pods using arguments, you cannot add a argument, you have to provide **all** the arguments if you want to add or change one.

Grafana Agents was chosen initially as a way to avoid a bulky Prometheus instance, as we only need the `remote write` functionality, so we can send metrics to Cortex.
```

<br>

### Reasons for switching back to Prometheus and Promtail

> - Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
>  It sounds like it would achive the same goal of giving us a instance that can `remote write` to cortex, without storing metrics locally.
> - Prometheus is more mature for the time being.
> - The prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing higher configuration options.
>   (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
> - For Promtail, I am currently seeing issues related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)

Reviewer 2: ```suggestion
- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
 It sounds like it would achieve the same goal of giving us a instance that can `remote write` to cortex without storing metrics locally.
- Prometheus is more mature for the time being.
- The Prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing more configuration options.
  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
- For Promtail, there is currently an issue related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)
```

### Recommendation

> So from my point of view we should consider going back and use prometheus and promtail,
> 
> This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.
> 
> I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki if we want to switch the logging agent (ex: to Grafana Agents when it is more mature).

Reviewer 2: ```suggestion
We should consider going back to using Prometheus and Promtail.

This is mainly due to the inconsistent documentation and early stage of the Grafana Agents, and we need something that is mature and has more consistent documentation.

The best option is to use the Prometheus operator for convenience, and then to deploy Promtail separately from Loki for extra flexibility. This way we do not need to change the way we deploy Loki if we want to switch the logging agent at a later stage (e.g. to Grafana Agents when it is more mature).
```

## Grafana agents - go/no go?

Reasons for not going with Grafana agents
- The documentation is confusing and the Agents and operator docs aren't aligened.

> - Some of the main sellingpoints (integrations and traces) aren't supported by the operator

Reviewer: ```suggestion
- Some of the main selling points (integrations and traces) aren't supported by the operator
   - We also expected one Grafana Agent per node for scrapping metrics to avoid a single-point-of-failure, but Grafana Agents aren't deployed that way
```

<br>

### Reasons for switching back to Prometheus and Promtail

- Prometheus has an experimental agent feature, and according to [this blog](https://prometheus.io/blog/2021/11/16/agent/) it has been tested at scale.
 It sounds like it would achive the same goal of giving us a instance that can `remote write` to cortex, without storing metrics locally.
- Prometheus is more mature for the time being.
- The prometheus operator could also be worth looking at since it give us some of the same features as the agents does, while still providing higher configuration options.
  (Also endorsment for prometheus-operator is higher - 6,7k stars vs 594 at the time of writing)
- For Promtail, I am currently seeing issues related to `filetarget.fsnotify.newwatcher: to many open files` which is said to be fixed by using `fs.inotify.max_user_instances` but the Grafana agents does not have a way of adding this to the logging pods, this promtail does [here](https://github.com/grafana/helm-charts/blob/e64169fab5fe248787bd62f4e1a990760fbf99a8/charts/promtail/values.yaml#L20)

> - In the current Agent configuration, we're using it as a package deal for Promtail and Prometheus.
>   - We send all logs to one place
>   - We want all metrics to be scraped by one agent anyways.

Reviewer: ```suggestion
```

### Recommendation

So from my point of view we should consider going back and use prometheus and promtail,

This is manily due to the inconsistent doc and early stage of the grafana agents, and we need something that is more mature and has more consisten doc.

> I would suggest using the prometheus operator for convinience and then deploying promtail seperate to Loki.

Reviewer: ```suggestion
I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki if we want to switch the logging agent (ex: to Grafana Agents when it is more mature).
```

> I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki we want to switch the logging agent (ex: to Grafana Agents when it is more mature).

Reviewer: ```suggestion
I would suggest using the prometheus operator for convinience and then deploying promtail separate to Loki for extra flexibility, so we don't need to touch Loki if we want to switch the logging agent (ex: to Grafana Agents when it is more mature).
```
